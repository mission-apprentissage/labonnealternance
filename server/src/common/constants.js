export const POURVUE = "Pourvue"
export const ANNULEE = "Annulée"
export const KEY_GENERATOR_PARAMS = ({ length, symbols, numbers }) => {
  return {
    length: length ?? 50,
    strict: true,
    numbers: numbers ?? true,
    symbols: symbols ?? true,
    lowercase: true,
    uppercase: false,
    excludeSimilarCharacters: true,
    exclude: '!"_%£$€*¨^=+~ß(){}[]§;,./:`@#&|<>?"',
  }
}
export const validation_utilisateur = {
  AUTO: "AUTOMATIQUE",
  MANUAL: "MANUELLE",
}
export const ENTREPRISE_DELEGATION = "ENTREPRISE_DELEGATION"
export const etat_utilisateur = {
  VALIDE: "VALIDÉ",
  DESACTIVE: "DESACTIVÉ",
  ATTENTE: "EN ATTENTE DE VALIDATION",
}
export const etat_etablissements = {
  ACTIF: "actif",
  FERME: "fermé",
}
export const ENTREPRISE = "ENTREPRISE"
export const CFA = "CFA"
export const REGEX = {
  SIRET: /^([0-9]{9}|[0-9]{14})$/,
  GEO: /^(-?\d+(\.\d+)?),\s*(-?\d+(\.\d+)?)$/,
  TELEPHONE: /^[0-9]{10}$/,
}
export const OPCOS = {
  AFDAS: "AFDAS",
  AKTO: "AKTO / Opco entreprises et salariés des services à forte intensité de main d'oeuvre",
  ATLAS: "ATLAS",
  CONSTRUCTYS: "Constructys",
  OPCOMMERCE: "L'Opcommerce",
  OCAPIAT: "OCAPIAT",
  OPCO2I: "OPCO 2i",
  EP: "Opco entreprises de proximité",
  MOBILITE: "Opco Mobilités",
  SANTE: "Opco Santé",
  UNIFORMATION: "Uniformation, l'Opco de la Cohésion sociale",
}
