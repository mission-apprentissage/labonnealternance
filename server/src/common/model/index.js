import mongoosePaginate from "mongoose-paginate"
import { getElasticInstance, mongoosastic } from "../esClient/index.js"
import schema from "../model/schema/index.js"
import { mongooseInstance } from "../mongodb.js"

const createModel = (modelName, descriptor, options = {}) => {
  const schema = new mongooseInstance.Schema(descriptor)

  if (options.esIndexName) {
    schema.plugin(mongoosastic, { esClient: getElasticInstance(), index: options.esIndexName })
  }

  if (options.esIndexName || options.paginate) {
    schema.plugin(mongoosePaginate)
  }

  if (options.createMongoDBIndexes) {
    options.createMongoDBIndexes(schema)
  }

  return mongooseInstance.model(modelName, schema)
}

const DomainesMetiers = createModel("domainesmetiers", schema.domainesMetiersSchema, {
  esIndexName: "domainesmetiers",
})
const FormationCatalogue = createModel("formationcatalogues", schema.mnaFormationSchema, {
  esIndexName: "formationcatalogues",
})
const DiplomesMetiers = createModel("diplomesmetiers", schema.diplomesMetiersSchema, {
  esIndexName: "diplomesmetiers",
})
const ApiCalls = createModel("apicalls", schema.apiCallSchema)
const Application = createModel("applications", schema.applicationSchema, {
  paginate: true,
})
const SourceFormations = createModel("sourceformations", schema.sourceFormationsSchema)
const GeoLocation = createModel("geolocation", schema.geoLocationSchema)
const EmailBlacklist = createModel("emailblacklist", schema.emailBlacklist)
const Opco = createModel("opco", schema.opco)
const BonnesBoites = createModel("bonnesboites", schema.bonneBoiteSchema, {
  esIndexName: "bonnesboites",
})
const User = createModel("user", schema.userSchema, { paginate: true })
const Appointment = createModel("appointment", schema.appointmentSchema, { paginate: true })
const WidgetParameter = createModel("widgetParameter", schema.widgetParameterSchema, { paginate: true })
const Etablissement = createModel("etablissement", schema.etablissementSchema, { paginate: true })
const ParcoursupEtablissementStat = createModel("parcoursupEtablissementStat", schema.parcoursupEtablissementStatSchema, { paginate: true })
const Formulaire = createModel("formulaire", schema.formulaireSchema, { esIndexName: "formulaires", paginate: true })
const Credential = createModel("credentials", schema.credentialSchema)
const Offre = createModel("offre", schema.offresSchema)
const Optout = createModel("optout", schema.optoutSchema, { paginate: true })
const ReferentielOpco = createModel("referentielOpco", schema.referentielOpcoSchema)
const UserRecruteur = createModel("userRecruteur", schema.userRecruteurSchema, { paginate: true })
const AppointmentDetailed = createModel("appointmentDetailed", schema.appointmentSchema)

export {
  DomainesMetiers,
  FormationCatalogue,
  DiplomesMetiers,
  ApiCalls,
  Application,
  SourceFormations,
  GeoLocation,
  EmailBlacklist,
  Opco,
  BonnesBoites,
  User,
  Appointment,
  WidgetParameter,
  Etablissement,
  ParcoursupEtablissementStat,
  Formulaire,
  Credential,
  Offre,
  Optout,
  ReferentielOpco,
  UserRecruteur,
  AppointmentDetailed,
}
