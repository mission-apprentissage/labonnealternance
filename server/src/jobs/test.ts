import { ObjectId } from "bson"
import { IApplicant, IApplication, ZApplicant } from "shared"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { asyncForEach } from "../common/utils/asyncUtils"

type ApplicationAggregate = Pick<IApplication, "applicant_email" | "applicant_first_name" | "applicant_last_name" | "applicant_phone" | "_id">

export const migrationTest = async () => {
  // console.log(extensions.telephone.safeParse("0690760552"))
  // return
  const applications = (await getDbCollection("applications")
    .find({}, { projection: { applicant_first_name: 1, applicant_last_name: 1, applicant_email: 1, applicant_phone: 1, _id: 1 } })
    .toArray()) as ApplicationAggregate[]

  const stat = { error: 0, success: 0, total: applications.length }

  await asyncForEach(applications, async (application) => {
    const { applicant_first_name, applicant_email, applicant_last_name, applicant_phone, _id } = application
    const now = new Date()
    const applicant: IApplicant = {
      _id: new ObjectId(),
      firstname: applicant_first_name,
      lastname: applicant_last_name,
      phone: applicant_phone,
      email: applicant_email,
      last_connection: null,
      createdAt: now,
      updatedAt: now,
    }
    const validation = ZApplicant.safeParse(applicant)
    if (!validation.success) {
      stat.error++
      console.log({ applicant, err: validation.error.errors })
    } else {
      stat.success++
    }
  })
  console.log(stat)
}
