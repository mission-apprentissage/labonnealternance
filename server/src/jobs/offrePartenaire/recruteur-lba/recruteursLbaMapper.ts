import { ObjectId } from "bson"
import proj4 from "proj4"
import { joinNonNullStrings } from "shared"
import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { IComputedJobsPartners } from "shared/models/jobsPartnersComputed.model"
import { IRecruteursLbaRaw } from "shared/models/rawRecruteursLba.model"

import { blankComputedJobPartner } from "../fillComputedJobsPartners"

export const recruteursLbaToJobPartners = (recruteursLba: IRecruteursLbaRaw): IComputedJobsPartners => {
  const {
    siret,
    enseigne,
    raison_sociale,
    activitePrincipaleEtablissement,
    labelActivitePrincipaleEtablissement,
    street_name,
    street_number,
    zip_code,
    email,
    phone,
    company_size,
    rome_codes,
    coordonneeLambertAbscisseEtablissement,
    coordonneeLambertOrdonneeEtablissement,
    libelleCommuneEtablissement,
    createdAt,
  } = recruteursLba

  return {
    ...blankComputedJobPartner(),
    _id: new ObjectId(),
    partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
    partner_job_id: siret,
    workplace_siret: siret,
    workplace_brand: enseigne,
    workplace_legal_name: raison_sociale,
    workplace_naf_code: activitePrincipaleEtablissement,
    workplace_naf_label: labelActivitePrincipaleEtablissement,
    workplace_address_city: libelleCommuneEtablissement,
    workplace_address_street_label: street_name,
    workplace_address_zipcode: zip_code,
    workplace_address_label: joinNonNullStrings([street_number, street_name, zip_code, libelleCommuneEtablissement]),
    workplace_geopoint: getWorkplaceGeolocation(coordonneeLambertAbscisseEtablissement, coordonneeLambertOrdonneeEtablissement, zip_code),
    workplace_size: company_size,
    apply_email: email,
    apply_phone: phone,
    offer_rome_codes: rome_codes.map(({ rome_code }) => rome_code),
    // laisser recruteurs_lba pour les entreprises issue de l'algorithme de lba (passer la validation & l'import dans jobs_partners)
    offer_creation: createdAt,
    offer_title: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
    offer_description: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
  }
}

// Définition des systèmes de coordonnées
const lambert93 = "EPSG:2154"
const laReunion = "EPSG:2975"
const guyane = "EPSG:2972"
const guadeloupeEtMartinique = "EPSG:5490"

const wgs84 = "EPSG:4326"

// Définition de Lambert 93 (au cas où proj4 ne l'aurait pas déjà en base)
proj4.defs(lambert93, "+proj=lcc +lat_1=49 +lat_2=44 +lat_0=46.5 +lon_0=3 " + "+x_0=700000 +y_0=6600000 +ellps=GRS80 +units=m +no_defs")
proj4.defs(laReunion, "+proj=utm +zone=40 +south +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs")
proj4.defs(guyane, "+proj=utm +zone=22 +south +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs")
proj4.defs(guadeloupeEtMartinique, "+proj=utm +zone=20 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs")

const getRegionFromZipCode = (zip_code: string | null): string => {
  if (!zip_code) return lambert93
  const prefix = zip_code.substring(0, 3)
  switch (prefix) {
    case "974":
      return laReunion
    case "971":
    case "972":
    case "977":
    case "978":
      return guadeloupeEtMartinique
    case "973":
      return guyane
    default:
      return lambert93
  }
}

export const getWorkplaceGeolocation = (x: number | null, y: number | null, zipCode: string | null): IComputedJobsPartners["workplace_geopoint"] => {
  if (x === 0 || y === 0 || x === null || y === null) return null

  const [longitude, latitude] = proj4(getRegionFromZipCode(zipCode), wgs84, [x, y])
  return { type: "Point", coordinates: [longitude, latitude] }
}
