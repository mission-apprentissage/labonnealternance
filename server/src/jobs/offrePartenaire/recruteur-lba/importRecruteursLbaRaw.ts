import { createReadStream, createWriteStream, existsSync, mkdirSync, unlinkSync } from "fs"
import { createRequire } from "module"
import { pipeline } from "node:stream/promises"
import path from "path"

import { internal } from "@hapi/boom"
import { ObjectId } from "bson"
import { groupData, oleoduc, transformData, writeData } from "oleoduc"
import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { IComputedJobsPartners } from "shared/models/jobsPartnersComputed.model"
import rawRecruteursLbaModel, { ZRecruteursLbaRaw } from "shared/models/rawRecruteursLba.model"

import __dirname from "@/common/dirname"
import { logger } from "@/common/logger"
import { getS3FileLastUpdate, s3ReadAsStream } from "@/common/utils/awsUtils"
import { getDbCollection } from "@/common/utils/mongodbUtils"
import { sentryCaptureException } from "@/common/utils/sentryUtils"
import { notifyToSlack } from "@/common/utils/slackUtils"
import config from "@/config"
import { isEmailBlacklisted } from "@/services/application.service"

import { recruteursLbaToJobPartners } from "./recruteursLbaMapper"

const require = createRequire(import.meta.url)

const { parser } = require("stream-json")
const { streamArray } = require("stream-json/streamers/StreamArray")

const CURRENT_DIR_PATH = __dirname(import.meta.url)
const PREDICTION_FILE_PATH = path.join(CURRENT_DIR_PATH, "./assets/recruteurslba.json")
const S3_FILE = config.algoRecuteursLba.s3File

export const createAssetsFolder = async () => {
  const assetsPath = path.join(CURRENT_DIR_PATH, "./assets")
  if (!(await existsSync(assetsPath))) {
    await mkdirSync(assetsPath)
  }
}

const removePredictionFile = async () => {
  try {
    logger.info("Deleting downloaded file from assets")
    await unlinkSync(PREDICTION_FILE_PATH)
  } catch (err) {
    logger.error("Error removing company algo file", err)
  }
}

export const checkIfAlgoFileAlreadyProcessed = async (): Promise<boolean> => {
  const algoFileLastModificationDate = await getS3FileLastUpdate("storage", S3_FILE)
  if (!algoFileLastModificationDate) {
    throw new Error("Aucune date de dernière modifications disponible sur le fichier issue de l'algo sur S3.")
  }

  const recruteurLbaRaw = await getDbCollection("raw_recruteurslba").findOne({}, { projection: { createdAt: 1 } })
  if (!recruteurLbaRaw) return false
  if (algoFileLastModificationDate.getTime() < recruteurLbaRaw.createdAt.getTime()) {
    await notifyToSlack({
      subject: `import des offres recruteurs lba dans raw`,
      message: `dernier fichier en date déjà traité.`,
    })
    return true
  }
  return false
}

const getRecruteursLbaFileFromS3 = async ({ from, to }) => {
  logger.info(`Downloading algo file from S3 Bucket`)
  await createAssetsFolder()
  const readStream = await s3ReadAsStream("storage", from)
  const writeStream = createWriteStream(to)

  try {
    await pipeline(readStream, writeStream)
    logger.info("File transfer completed successfully")
  } catch (error) {
    logger.error("Error during file transfer:", error)
    throw error
  }
}

const importRecruteursLbaToRawCollection = async () => {
  logger.info("deleting old data")
  await getDbCollection("raw_recruteurslba").deleteMany({})
  logger.info("import starting...")
  let count = 0
  const now = new Date()
  await oleoduc(
    createReadStream(PREDICTION_FILE_PATH),
    parser(),
    streamArray(),
    transformData((doc) => {
      const recruteur = { createdAt: now, _id: new ObjectId(), ...doc.value }
      if (!ZRecruteursLbaRaw.safeParse(recruteur).success) return null
      return recruteur
    }),
    groupData({ size: 10_000 }),
    writeData((array) => {
      const filtered = array.filter((item) => item)
      if (!filtered.length) return
      count += filtered.length
      getDbCollection("raw_recruteurslba").insertMany(filtered)
    })
  )
  const message = `import recruteurs lba terminé : ${count} recruteurs importées`
  logger.info(message)
  notifyToSlack({
    subject: `import des offres recruteurs lba dans raw`,
    message,
  })
}

export const importRecruteursLbaRaw = async () => {
  try {
    await getRecruteursLbaFileFromS3({ from: S3_FILE, to: PREDICTION_FILE_PATH })
    await importRecruteursLbaToRawCollection()
  } catch (err) {
    await notifyToSlack({ subject: `import des offres recruteurs lba dans raw`, message: `import recruteurs lba terminé : echec de l'import`, error: true })
    logger.error(err)
    sentryCaptureException(err)
  } finally {
    await removePredictionFile()
  }
}

export const importRecruteurLbaToComputed = async () => {
  const collectionSource = rawRecruteursLbaModel.collectionName
  const partnerLabel = JOBPARTNERS_LABEL.RECRUTEURS_LBA
  const zodInput = ZRecruteursLbaRaw
  const mapper = recruteursLbaToJobPartners

  logger.info(`début d'import dans computed_jobs_partners pour partner_label=${partnerLabel}`)
  const counters = { total: 0, success: 0, error: 0 }
  const importDate = new Date()
  await oleoduc(
    getDbCollection(collectionSource).find({}).stream(),
    writeData(
      async (document: any) => {
        counters.total++
        try {
          const parsedDocument = zodInput.parse(document)
          const { apply_email, apply_phone, updated_at, ...rest } = mapper(parsedDocument)

          const isEmailBl = apply_email ? await isEmailBlacklisted(apply_email) : false

          await getDbCollection("computed_jobs_partners").updateOne(
            { workplace_siret: rest.workplace_siret },
            {
              $set: { apply_email: isEmailBl ? null : apply_email, apply_phone, updated_at: importDate },
              $setOnInsert: { ...rest, offer_status_history: [], _id: new ObjectId() },
            },
            {
              upsert: true,
            }
          )
          counters.success++
        } catch (err) {
          counters.error++
          const newError = internal(`error converting raw job to partner_label job for id=${document._id} partner_label=${partnerLabel}`)
          logger.error(newError.message, err)
          newError.cause = err
          sentryCaptureException(newError)
        }
      },
      { parallel: 10 }
    )
  )
  const message = `import dans computed_jobs_partners pour partner_label=${partnerLabel} terminé. total=${counters.total}, success=${counters.success}, errors=${counters.error}`
  logger.info(message)
  await notifyToSlack({
    subject: `mapping Raw => computed_jobs_partners`,
    message,
    error: counters.error > 0,
  })
}

export const removeMissingRecruteursLbaFromRaw = async () => {
  logger.info("clean-up recruteurs_lba in computed_jobs_partners from raw")
  const results = (await getDbCollection("computed_jobs_partners")
    .aggregate([
      { $match: { partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA } },
      {
        $lookup: {
          from: "raw_recruteurslba",
          localField: "workplace_siret",
          foreignField: "siret",
          as: "matching",
        },
      },
      { $match: { matching: { $size: 0 } } },
      { $project: { _id: 1 } },
    ])
    .toArray()) as IComputedJobsPartners[]

  const idsToRemove = results.map(({ _id }) => _id)

  if (idsToRemove.length) {
    await getDbCollection("computed_jobs_partners").deleteMany({ _id: { $in: idsToRemove } })
  }
  const message = `clean-up dans computed_jobs_partners pour partner_label=${JOBPARTNERS_LABEL.RECRUTEURS_LBA} terminé. total=${idsToRemove.length}`
  logger.info(message)
  await notifyToSlack({
    subject: `mapping Raw => computed_jobs_partners`,
    message,
  })
}
