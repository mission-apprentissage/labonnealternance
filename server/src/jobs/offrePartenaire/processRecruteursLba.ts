/**
 * Recruteurs LBA alorithm file is updated every sunday (normally)
 * This file is processed every sunday
 *
 */
import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"

import { getDbCollection } from "../../common/utils/mongodbUtils"

import { fillComputedJobsPartners } from "./fillComputedJobsPartners"
import { importFromComputedToJobsPartners } from "./importFromComputedToJobsPartners"
import { checkIfAlgoFileAlreadyProcessed, importRecruteurLbaToComputed, importRecruteursLbaRaw } from "./recruteur-lba/importRecruteursLbaRaw"

const filter = {
  partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
}

export const processRecruteursLba = async () => {
  const fileAlreadyProcessed = await checkIfAlgoFileAlreadyProcessed()
  if (fileAlreadyProcessed) return

  await importRecruteursLbaRaw()
  await importRecruteurLbaToComputed()
  await fillComputedJobsPartners(filter)
  await getDbCollection("jobs_partners").deleteMany(filter)
  await importFromComputedToJobsPartners(filter)
}
