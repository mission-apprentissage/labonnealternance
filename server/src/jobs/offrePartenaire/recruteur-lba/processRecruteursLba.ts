/**
 * Recruteurs LBA alorithm file is updated every sunday (normally)
 * This file is processed every sunday
 *
 */

import Stream from "node:stream"

import { Filter } from "mongodb"
import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { IComputedJobsPartners } from "shared/models/jobsPartnersComputed.model"

import { logger } from "@/common/logger"
import { cancelRemovedJobsPartners } from "@/jobs/offrePartenaire/cancelRemovedJobsPartners"
import { fillComputedRecruteursLba } from "@/jobs/offrePartenaire/fillComputedRecruteursLba"
import { importFromComputedToJobsPartners } from "@/jobs/offrePartenaire/importFromComputedToJobsPartners"

import { checkIfAlgoFileAlreadyProcessed, importRecruteurLbaToComputed, importRecruteursLbaRaw } from "./importRecruteursLbaRaw"

export const processRecruteursLba = async ({ sourceFileReadStream, skipCheckFileDate = false }: { sourceFileReadStream?: Stream.Readable; skipCheckFileDate?: boolean } = {}) => {
  logger.info("début de processRecruteursLba")
  if (!skipCheckFileDate) {
    const fileAlreadyProcessed = await checkIfAlgoFileAlreadyProcessed()
    if (fileAlreadyProcessed) {
      logger.info("processRecruteursLba: le fichier a déjà été traité")
      return
    }
  }

  await importRecruteursLbaRaw(sourceFileReadStream)
  await importRecruteurLbaToComputed()
  await fillComputedRecruteursLba()

  const filter: Filter<IComputedJobsPartners> = {
    partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
  }
  await importFromComputedToJobsPartners(filter)
  await cancelRemovedJobsPartners(filter)
  logger.info("fin de processRecruteursLba")
}
