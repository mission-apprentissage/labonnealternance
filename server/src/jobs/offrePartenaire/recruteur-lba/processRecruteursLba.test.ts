import fs from "node:fs"

import omit from "lodash-es/omit"
import { ObjectId } from "mongodb"
import { OPCOS_LABEL } from "shared/constants/recruteur"
import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { IRecruteursLbaRaw } from "shared/models/rawRecruteursLba.model"
import { beforeEach, describe, expect, it, vi } from "vitest"

import { getDbCollection } from "@/common/utils/mongodbUtils"
import { stringToStream } from "@/common/utils/streamUtils"
import { processRecruteursLba } from "@/jobs/offrePartenaire/recruteur-lba/processRecruteursLba"
import { createJobPartner } from "@tests/utils/jobsPartners.test.utils"
import { useMongo } from "@tests/utils/mongo.test.utils"

import { generateRecruiterRawFixture } from "../../../../../shared/src/fixtures/recruiterRaw.fixture"

useMongo()

const now = new Date("2024-07-21T04:49:06.000+02:00")

const testDir = "server/src/jobs/offrePartenaire/recruteur-lba/__tests__"

const processRecruteursLbaFromFile = async (sourceFilePath: string) => {
  const sourceFileReadStream = fs.createReadStream(sourceFilePath)
  await processRecruteursLba({ skipCheckFileDate: true, sourceFileReadStream })
}

const processRecruteursLbaFromObjects = async (recruteursRaw: IRecruteursLbaRaw[]) => {
  const stream = stringToStream(JSON.stringify(recruteursRaw))
  await processRecruteursLba({ skipCheckFileDate: true, sourceFileReadStream: stream })
}

describe("importRecruteursLbaRaw", () => {
  beforeEach(async () => {
    vi.useFakeTimers()
    vi.setSystemTime(now)

    await getDbCollection("opcos").insertMany([
      {
        _id: new ObjectId(),
        siren: "808413827",
        opco: OPCOS_LABEL.ATLAS,
        idcc: 1000,
      },
      {
        _id: new ObjectId(),
        siren: "130025265",
        opco: OPCOS_LABEL.ATLAS,
        idcc: 1000,
      },
    ])

    return async () => {
      vi.useRealTimers()
      await getDbCollection("opcos").deleteMany({})
      await getDbCollection("unsubscribedrecruteurslba").deleteMany({})
      await getDbCollection("emailblacklists").deleteMany({})

      await getDbCollection("raw_recruteurslba").deleteMany({})
      await getDbCollection("computed_jobs_partners").deleteMany({})
      await getDbCollection("jobs_partners").deleteMany({})
    }
  })

  it("should process all recruteurs", async () => {
    await processRecruteursLbaFromFile(`${testDir}/processRecruteursLba.test.1.json`)
    const publishedJobPartners = await getDbCollection("jobs_partners").find({ partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA }).toArray()
    expect.soft(publishedJobPartners.map((job) => omit(job, "_id"))).toMatchSnapshot()
  })

  it("should remove existing recruteurs in job_partners that are not in the file", async () => {
    // given
    await createJobPartner({
      partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA,
    })
    expect.soft((await getDbCollection("jobs_partners").find({ partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA }).toArray()).length).toBe(1)

    // when
    await processRecruteursLbaFromFile(`${testDir}/processRecruteursLba.test.1.json`)
    // then
    const publishedJobPartners = await getDbCollection("jobs_partners").find({ partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA }).toArray()
    expect.soft(publishedJobPartners.length).toBe(2)
  })
  it("should remove email contact when it is blacklisted", async () => {
    // given
    const email = "roget@dinum.fr"
    await getDbCollection("emailblacklists").insertOne({
      _id: new ObjectId(),
      created_at: new Date(),
      email,
      blacklisting_origin: "test",
    })

    // when
    await processRecruteursLbaFromObjects([generateRecruiterRawFixture({ email })])
    // then
    const publishedJobPartners = await getDbCollection("jobs_partners").find({ partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA }).toArray()
    expect.soft(publishedJobPartners.length).toBe(1)
    const [job] = publishedJobPartners
    expect.soft(job.apply_email).toBe(null)
  })
  it("should not import recruiters that opted out", async () => {
    // given
    const siret = "13002526500013"
    await getDbCollection("unsubscribedrecruteurslba").insertOne({
      _id: new ObjectId(),
      siret,
      raison_sociale: "DINUM",
      enseigne: "",
      naf_code: "",
      naf_label: "",
      rome_codes: [],
      insee_city_code: null,
      zip_code: null,
      city: null,
      company_size: null,
      created_at: new Date(),
      last_update_at: new Date(),
      unsubscribe_date: new Date(),
      unsubscribe_reason: "raison",
    })

    // when
    await processRecruteursLbaFromObjects([generateRecruiterRawFixture({ siret })])
    // then
    const publishedJobPartners = await getDbCollection("jobs_partners").find({ partner_label: JOBPARTNERS_LABEL.RECRUTEURS_LBA }).toArray()
    expect.soft(publishedJobPartners.length).toBe(0)
  })
})
