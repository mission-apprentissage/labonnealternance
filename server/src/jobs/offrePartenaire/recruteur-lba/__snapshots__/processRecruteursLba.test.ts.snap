// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importRecruteursLbaRaw > should process all recruteurs 1`] = `
[
  {
    "apply_email": null,
    "apply_phone": null,
    "apply_url": null,
    "contract_duration": null,
    "contract_is_disabled_elligible": false,
    "contract_remote": null,
    "contract_start": null,
    "contract_type": [],
    "created_at": 2024-07-21T02:49:06.000Z,
    "offer_access_conditions": [],
    "offer_creation": 2024-07-21T02:49:06.000Z,
    "offer_description": "recruteurs_lba",
    "offer_desired_skills": [],
    "offer_expiration": null,
    "offer_multicast": false,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": [
      "D1208",
      "I1203",
    ],
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": null,
    "offer_title": "recruteurs_lba",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "80841382700011",
    "partner_label": "recruteurs_lba",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "workplace_address_city": "LIVERDUN",
    "workplace_address_label": "54460 LIVERDUN",
    "workplace_address_street_label": null,
    "workplace_address_zipcode": "54460",
    "workplace_brand": "DRYSOFT",
    "workplace_description": null,
    "workplace_geopoint": {
      "coordinates": [
        49.90490699998852,
        1.105230000000001,
      ],
      "type": "Point",
    },
    "workplace_idcc": 1000,
    "workplace_legal_name": "ETABLISSEMENTS FRERES",
    "workplace_naf_code": "5510Z",
    "workplace_naf_label": "Hôtels et hébergement similaire ",
    "workplace_name": null,
    "workplace_opco": "ATLAS",
    "workplace_siret": "80841382700011",
    "workplace_size": "1-2",
    "workplace_website": null,
  },
  {
    "apply_email": "roget@dinum.fr",
    "apply_phone": "0123456789",
    "apply_url": null,
    "contract_duration": null,
    "contract_is_disabled_elligible": false,
    "contract_remote": null,
    "contract_start": null,
    "contract_type": [],
    "created_at": 2024-07-21T02:49:06.000Z,
    "offer_access_conditions": [],
    "offer_creation": 2024-07-21T02:49:06.000Z,
    "offer_description": "recruteurs_lba",
    "offer_desired_skills": [],
    "offer_expiration": null,
    "offer_multicast": false,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": [
      "M1607",
      "M1203",
    ],
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": null,
    "offer_title": "recruteurs_lba",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "13002526500013",
    "partner_label": "recruteurs_lba",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "workplace_address_city": "SAINT-ROMAIN-DE-LERPS",
    "workplace_address_label": "CHEMIN DU BALAYOU 07130 SAINT-ROMAIN-DE-LERPS",
    "workplace_address_street_label": "CHEMIN DU BALAYOU",
    "workplace_address_zipcode": "07130",
    "workplace_brand": "ETABLISSEMENTS DINUM",
    "workplace_description": null,
    "workplace_geopoint": {
      "coordinates": [
        49.90490699998852,
        1.105230000000001,
      ],
      "type": "Point",
    },
    "workplace_idcc": 1000,
    "workplace_legal_name": "ETABLISSEMENTS DINUM",
    "workplace_naf_code": "4673A",
    "workplace_naf_label": "Commerce de gros (commerce interentreprises) de bois et de matériaux de construction ",
    "workplace_name": null,
    "workplace_opco": "ATLAS",
    "workplace_siret": "13002526500013",
    "workplace_size": "6-9",
    "workplace_website": null,
  },
]
`;
